Sources/Extraction/Filters/HillOrientation.c

```diff
@@ -1,26 +1,183 @@
+        {
+            if (!BinaryMap_GetBit(blockMask, blockX, blockY))
+                continue;
 
-    return ho;
+            int bottomLeftX = blocks->blockAreas.corners.allX.data[blockX];
+            int bottomLeftY = blocks->blockAreas.corners.allY.data[blockY];
+            for (int x = bottomLeftX; x < bottomLeftX + blockWidth; ++x) {
+                for ( int y = bottomLeftY; y < bottomLeftY + blockHeight; ++y ) {
+                    blockOrientations.data[blockX][blockY] = Calc_Add2PointsF(&blockOrientations.data[blockX][blockY], &pixelOrientations->data[x][y]);
+                }
+            }
+        }
+    }
+    return blockOrientations;
 }
 
-void HillOrientation_Destruct(HillOrientation *me)
+PointFArray2D HillOrientation_SmoothDirections(PointFArray2D directions, BinaryMap blockMask)
 {
+    // PointFArray2D smoothedDirections = PointFArray2D_Construct()
+    return  (PointFArray2D) { NULL };
 }
 
-void /*UInt8Array2D*/ HillOrientation_Detect(const HillOrientation *me, const FloatArray2D *image, const BinaryMap *mask, const BlockMap *blocks)
+BoolArray2D HillOrientation_BlockMapToPixelMask(Size imageDimensions, BinaryMap * blockMask, BlockMap * blocks) {
+    const int blockHeight = 15; //blocks->pixelCount.height / blocks->blockCount.height;
+    const int blockWidth = 15; //blocks->pixelCount.width / blocks->blockCount.width;
+
+    BoolArray2D pixelMask = BoolArray2D_Construct(imageDimensions.width, imageDimensions.height);
+
+    for (int blockX = 0; blockX < blocks->blockCount.width; blockX++)
+    {
+        for (int blockY = 0; blockY < blocks->blockCount.height; blockY++)
+        {
+            bool blockIsInImage = BinaryMap_GetBit(blockMask, blockX, blockY);
+            int bottomLeftX = blocks->blockAreas.corners.allX.data[blockX];
+            int bottomLeftY = blocks->blockAreas.corners.allY.data[blockY];
+            for (int x = bottomLeftX; x < bottomLeftX + blockWidth; ++x) {
+                for ( int y = bottomLeftY; y < bottomLeftY + blockHeight; ++y ) {
+                    pixelMask.data[x][y] = blockIsInImage;
+                }
+            }
+        }
+    }
+    return pixelMask;
+}
+
+PointFArray2D HillOrientation_Smooth(PointFArray2D directions, BinaryMap * mask)
 {
-    /* TODO: Implement HillOrientation_Detect() */
+    return (PointFArray2D) { NULL };
+}
+
+UInt16Array2D HillOrientation_DirectionsToAngles(PointFArray2D directions, BinaryMap *mask)
+{
+    UInt16Array2D angles = UInt16Array2D_Construct(mask->width, mask->height);
+
+    for(int x = 0; x < directions.sizeX; x++)
+    {
+        for(int y = 0; y < directions.sizeY; y++)
+        {
+            if(!BinaryMap_GetBit(mask, x, y))
+                continue;
+
+            double angle = atan2(directions.data[x][y].y, directions.data[x][y].x);
+            //multiply by 2 to collapse opposite angles onto each other
+            angles.data[x][y] = (uint16_t)(angle / 3.1415f * 360 * 2 + 90) % 360;
```

Comment:
I tried an angle or 45 & 225 
and 90/270 
and I could not make it give me the expected answer
I think that Radians-> Degrees is (radians/PI) * 180
So the calc I was checking was (degrees * 2 * 2) % 360 ie ignoring the add 90degrees
Cheers David